diff -u -N -r linux-3.18.1.orig/include/linux/ethtool.h linux-3.18.1/include/linux/ethtool.h
--- linux-3.18.1.orig/include/linux/ethtool.h	2014-12-16 17:39:45.000000000 +0000
+++ linux-3.18.1/include/linux/ethtool.h	2014-12-27 00:06:16.621624475 +0000
@@ -40,8 +40,14 @@
 
 #include <linux/rculist.h>
 
+#ifdef CONFIG_NET_ETHTOOL
 extern int __ethtool_get_settings(struct net_device *dev,
 				  struct ethtool_cmd *cmd);
+#else
+static inline int __ethtool_get_settings(struct net_device *dev,
+					 struct ethtool_cmd *cmd)
+{ return -EINVAL; }
+#endif
 
 /**
  * enum ethtool_phys_id_state - indicator state for physical identification
@@ -61,9 +67,17 @@
 
 struct net_device;
 
+#ifdef CONFIG_NET_ETHTOOL
 /* Some generic methods drivers may use in their ethtool_ops */
 u32 ethtool_op_get_link(struct net_device *dev);
 int ethtool_op_get_ts_info(struct net_device *dev, struct ethtool_ts_info *eti);
+#else
+/* Some generic methods drivers may use in their ethtool_ops */
+static inline u32 ethtool_op_get_link(struct net_device *dev) { return 0; }
+static inline int
+ethtool_op_get_ts_info(struct net_device *dev, struct ethtool_ts_info *eti)
+{ return -EINVAL; }
+#endif
 
 /**
  * ethtool_rxfh_indir_default - get default value for RX flow hash indirection
diff -u -N -r linux-3.18.1.orig/include/linux/netdevice.h linux-3.18.1/include/linux/netdevice.h
--- linux-3.18.1.orig/include/linux/netdevice.h	2014-12-16 17:39:45.000000000 +0000
+++ linux-3.18.1/include/linux/netdevice.h	2014-12-27 00:09:06.545617742 +0000
@@ -58,8 +58,25 @@
 /* 802.11 specific */
 struct wireless_dev;
 
+#ifdef CONFIG_NET_ETHTOOL
+/* When the driver uses this macro ethtool code can be optimized out
+ * when not needed. We still reference it to avoid unused static
+ * warnings.
+ */
+#define SET_ETHTOOL_OPS(netdev,ops) \
+       ( (void)(ops), (netdev)->ethtool_ops = (ops) )
+
 void netdev_set_default_ethtool_ops(struct net_device *dev,
 				    const struct ethtool_ops *ops);
+int dev_ethtool(struct net *net, struct ifreq *);
+#else
+#define SET_ETHTOOL_OPS(netdev,ops) do {} while(0)
+static inline void
+netdev_set_default_ethtool_ops(struct net_device *dev,
+                              const struct ethtool_ops *ops) {}
+static inline int
+dev_ethtool(struct net *net, struct ifreq *ifr) { return -EINVAL; }
+#endif
 
 /* Backlog congestion levels */
 #define NET_RX_SUCCESS		0	/* keep 'em coming, baby */
@@ -2850,7 +2867,6 @@
 
 bool dev_valid_name(const char *name);
 int dev_ioctl(struct net *net, unsigned int cmd, void __user *);
-int dev_ethtool(struct net *net, struct ifreq *);
 unsigned int dev_get_flags(const struct net_device *);
 int __dev_change_flags(struct net_device *, unsigned int flags);
 int dev_change_flags(struct net_device *, unsigned int);
diff -u -N -r linux-3.18.1.orig/net/core/dev.c linux-3.18.1/net/core/dev.c
--- linux-3.18.1.orig/net/core/dev.c	2014-12-16 17:39:45.000000000 +0000
+++ linux-3.18.1/net/core/dev.c	2014-12-27 00:11:02.297613156 +0000
@@ -6574,6 +6574,7 @@
 
 static const struct ethtool_ops default_ethtool_ops;
 
+#ifdef CONFIG_NET_ETHTOOL
 void netdev_set_default_ethtool_ops(struct net_device *dev,
 				    const struct ethtool_ops *ops)
 {
@@ -6581,6 +6582,7 @@
 		dev->ethtool_ops = ops;
 }
 EXPORT_SYMBOL_GPL(netdev_set_default_ethtool_ops);
+#endif
 
 void netdev_freemem(struct net_device *dev)
 {
diff -u -N -r linux-3.18.1.orig/net/core/Makefile linux-3.18.1/net/core/Makefile
--- linux-3.18.1.orig/net/core/Makefile	2014-12-16 17:39:45.000000000 +0000
+++ linux-3.18.1/net/core/Makefile	2014-12-27 00:10:25.613614609 +0000
@@ -7,12 +7,13 @@
 
 obj-$(CONFIG_SYSCTL) += sysctl_net_core.o
 
-obj-y		     += dev.o ethtool.o dev_addr_lists.o dst.o netevent.o \
+obj-y		     += dev.o dev_addr_lists.o dst.o netevent.o \
 			neighbour.o rtnetlink.o utils.o link_watch.o filter.o \
 			sock_diag.o dev_ioctl.o tso.o
 
 obj-$(CONFIG_XFRM) += flow.o
 obj-y += net-sysfs.o
+obj-$(CONFIG_NET_ETHTOOL) += ethtool.o
 obj-$(CONFIG_PROC_FS) += net-procfs.o
 obj-$(CONFIG_NET_PKTGEN) += pktgen.o
 obj-$(CONFIG_NETPOLL) += netpoll.o
diff -u -N -r linux-3.18.1.orig/net/Kconfig linux-3.18.1/net/Kconfig
--- linux-3.18.1.orig/net/Kconfig	2014-12-16 17:39:45.000000000 +0000
+++ linux-3.18.1/net/Kconfig	2014-12-27 00:09:49.177616053 +0000
@@ -83,6 +83,12 @@
 
 endif # if INET
 
+config NET_ETHTOOL
+	bool "Ethtool support"
+	default y
+	help
+	  Support changing ethernet driver parameters from user tools.
+
 config NETWORK_SECMARK
 	bool "Security Marking"
 	help
