diff -u -r -N linux-4.0.4.orig/include/linux/ethtool.h linux-4.0.4/include/linux/ethtool.h
--- linux-4.0.4.orig/include/linux/ethtool.h	2015-06-07 17:53:26.897238807 +0100
+++ linux-4.0.4/include/linux/ethtool.h	2015-06-07 17:53:20.961239010 +0100
@@ -40,8 +40,14 @@
 
 #include <linux/rculist.h>
 
+#ifdef CONFIG_NET_ETHTOOL
 extern int __ethtool_get_settings(struct net_device *dev,
 				  struct ethtool_cmd *cmd);
+#else
+static inline int __ethtool_get_settings(struct net_device *dev,
+					 struct ethtool_cmd *cmd)
+{ return -EINVAL; }
+#endif
 
 /**
  * enum ethtool_phys_id_state - indicator state for physical identification
@@ -81,9 +87,17 @@
 
 struct net_device;
 
+#ifdef CONFIG_NET_ETHTOOL
 /* Some generic methods drivers may use in their ethtool_ops */
 u32 ethtool_op_get_link(struct net_device *dev);
 int ethtool_op_get_ts_info(struct net_device *dev, struct ethtool_ts_info *eti);
+#else
+/* Some generic methods drivers may use in their ethtool_ops */
+static inline u32 ethtool_op_get_link(struct net_device *dev) { return 0; }
+static inline int
+ethtool_op_get_ts_info(struct net_device *dev, struct ethtool_ts_info *eti)
+{ return -EINVAL; }
+#endif
 
 /**
  * ethtool_rxfh_indir_default - get default value for RX flow hash indirection
diff -u -r -N linux-4.0.4.orig/include/linux/netdevice.h linux-4.0.4/include/linux/netdevice.h
--- linux-4.0.4.orig/include/linux/netdevice.h	2015-06-07 17:53:26.901238807 +0100
+++ linux-4.0.4/include/linux/netdevice.h	2015-06-07 17:53:20.965239010 +0100
@@ -61,8 +61,25 @@
 /* 802.15.4 specific */
 struct wpan_dev;
 
+#ifdef CONFIG_NET_ETHTOOL
+/* When the driver uses this macro ethtool code can be optimized out
+ * when not needed. We still reference it to avoid unused static
+ * warnings.
+ */
+#define SET_ETHTOOL_OPS(netdev,ops) \
+       ( (void)(ops), (netdev)->ethtool_ops = (ops) )
+
 void netdev_set_default_ethtool_ops(struct net_device *dev,
 				    const struct ethtool_ops *ops);
+int dev_ethtool(struct net *net, struct ifreq *);
+#else
+#define SET_ETHTOOL_OPS(netdev,ops) do {} while(0)
+static inline void
+netdev_set_default_ethtool_ops(struct net_device *dev,
+                              const struct ethtool_ops *ops) {}
+static inline int
+dev_ethtool(struct net *net, struct ifreq *ifr) { return -EINVAL; }
+#endif
 
 /* Backlog congestion levels */
 #define NET_RX_SUCCESS		0	/* keep 'em coming, baby */
@@ -2960,7 +2977,6 @@
 
 bool dev_valid_name(const char *name);
 int dev_ioctl(struct net *net, unsigned int cmd, void __user *);
-int dev_ethtool(struct net *net, struct ifreq *);
 unsigned int dev_get_flags(const struct net_device *);
 int __dev_change_flags(struct net_device *, unsigned int flags);
 int dev_change_flags(struct net_device *, unsigned int);
diff -u -r -N linux-4.0.4.orig/net/core/dev.c linux-4.0.4/net/core/dev.c
--- linux-4.0.4.orig/net/core/dev.c	2015-06-07 17:53:26.901238807 +0100
+++ linux-4.0.4/net/core/dev.c	2015-06-07 17:53:20.965239010 +0100
@@ -6709,6 +6709,7 @@
 
 static const struct ethtool_ops default_ethtool_ops;
 
+#ifdef CONFIG_NET_ETHTOOL
 void netdev_set_default_ethtool_ops(struct net_device *dev,
 				    const struct ethtool_ops *ops)
 {
@@ -6716,6 +6717,7 @@
 		SET_ETHTOOL_OPS(dev, ops);
 }
 EXPORT_SYMBOL_GPL(netdev_set_default_ethtool_ops);
+#endif
 
 void netdev_freemem(struct net_device *dev)
 {
diff -u -r -N linux-4.0.4.orig/net/core/Makefile linux-4.0.4/net/core/Makefile
--- linux-4.0.4.orig/net/core/Makefile	2015-06-07 17:53:26.901238807 +0100
+++ linux-4.0.4/net/core/Makefile	2015-06-07 17:53:20.965239010 +0100
@@ -8,13 +8,14 @@
 
 obj-$(CONFIG_SYSCTL) += sysctl_net_core.o
 
-obj-y		     += dev.o ethtool.o dev_addr_lists.o dst.o netevent.o \
+obj-y		     += dev.o dev_addr_lists.o dst.o netevent.o \
 			neighbour.o rtnetlink.o utils.o link_watch.o filter.o \
 			dev_ioctl.o tso.o
 
 obj-$(CONFIG_SOCK_DIAG) += sock_diag.o
 obj-$(CONFIG_XFRM) += flow.o
 obj-y += net-sysfs.o
+obj-$(CONFIG_NET_ETHTOOL) += ethtool.o
 obj-$(CONFIG_PROC_FS) += net-procfs.o
 obj-$(CONFIG_NET_PKTGEN) += pktgen.o
 obj-$(CONFIG_LPF_FILTER) += filter.o
diff -u -r -N linux-4.0.4.orig/net/core/sysctl_net_core.c linux-4.0.4/net/core/sysctl_net_core.c
--- linux-4.0.4.orig/net/core/sysctl_net_core.c	2015-05-17 17:55:21.000000000 +0100
+++ linux-4.0.4/net/core/sysctl_net_core.c	2015-06-07 17:54:04.057237539 +0100
@@ -279,6 +279,7 @@
 		.mode		= 0644,
 		.proc_handler	= proc_dointvec
 	},
+#ifdef CONFIG_NET_ETHTOOL
 	{
 		.procname	= "netdev_rss_key",
 		.data		= &netdev_rss_key,
@@ -286,6 +287,7 @@
 		.mode		= 0444,
 		.proc_handler	= proc_do_rss_key,
 	},
+#endif
 #ifdef CONFIG_BPF_JIT
 	{
 		.procname	= "bpf_jit_enable",
diff -u -r -N linux-4.0.4.orig/net/Kconfig linux-4.0.4/net/Kconfig
--- linux-4.0.4.orig/net/Kconfig	2015-06-07 17:53:26.901238807 +0100
+++ linux-4.0.4/net/Kconfig	2015-06-07 17:53:20.965239010 +0100
@@ -98,6 +98,12 @@
 config SOCK_DIAG
 	bool
 
+config NET_ETHTOOL
+	bool "Ethtool support"
+	default y
+	help
+	  Support changing ethernet driver parameters from user tools.
+
 config NETWORK_SECMARK
 	bool "Security Marking"
 	help
