--- linux-3.18.1/arch/mips/ar7/platform.c.orig	2014-12-29 21:30:28.930740854 +0000
+++ linux-3.18.1/arch/mips/ar7/platform.c	2014-12-29 21:31:31.534738373 +0000
@@ -581,7 +581,7 @@
 	uart_port.type		= PORT_AR7;
 	uart_port.uartclk	= clk_get_rate(bus_clk) / 2;
 	uart_port.iotype	= UPIO_MEM32;
-	uart_port.flags		= UPF_FIXED_TYPE;
+	uart_port.flags		= UPF_FIXED_PORT | UPF_FIXED_TYPE | UPF_BOOT_AUTOCONF;
 	uart_port.regshift	= 2;
 
 	uart_port.line		= 0;
