--- a/arch/mips/lantiq/xway/ath_eep.c	2015-08-23 11:45:21.693976108 +0100
+++ b/arch/mips/lantiq/xway/ath_eep.c	2015-08-23 11:47:15.845971898 +0100
@@ -14,8 +14,12 @@
 #include <linux/module.h>
 #include <linux/platform_device.h>
 #include <linux/etherdevice.h>
+#ifdef CONFIG_ATH5K
 #include <linux/ath5k_platform.h>
+#endif
+#ifdef CONFIG_ATH9K
 #include <linux/ath9k_platform.h>
+#endif
 #include <linux/pci.h>
 #include <linux/err.h>
 #include <linux/mtd/mtd.h>
@@ -23,7 +27,8 @@
 #include <lantiq_soc.h>
 
 extern int (*ltq_pci_plat_dev_init)(struct pci_dev *dev);
-struct ath5k_platform_data ath5k_pdata;
+
+#ifdef CONFIG_ATH5K
 struct ath9k_platform_data ath9k_pdata = {
 	.led_pin = -1,
 };
@@ -174,7 +179,10 @@
 }
 late_initcall(of_ath9k_eeprom_init_late);
 subsys_initcall(of_ath9k_eeprom_init);
+#endif
 
+#ifdef CONFIG_ATH5K
+struct ath5k_platform_data ath5k_pdata;
 
 static int ath5k_pci_plat_dev_init(struct pci_dev *dev)
 {
@@ -280,3 +288,4 @@
 	return platform_driver_probe(&ath5k_eeprom_driver, of_ath5k_eeprom_probe);
 }
 device_initcall(of_ath5k_eeprom_init);
+#endif
--- a/arch/mips/lantiq/xway/Makefile	2015-08-23 11:50:23.925964962 +0100
+++ b/arch/mips/lantiq/xway/Makefile	2015-08-23 11:50:47.489964093 +0100
@@ -3,6 +3,6 @@
 obj-y += vmmc.o tffs.o
 
 obj-y += eth_mac.o
-obj-$(CONFIG_PCI) += ath_eep.o rt_eep.o pci-ath-fixup.o
+obj-$(CONFIG_PCI) += ath_eep.o pci-ath-fixup.o
 
 obj-$(CONFIG_XRX200_PHY_FW) += xrx200_phy_fw.o
